Automattic/wp-calypso
# composite-checkout: Combine form and transaction status state

## Proposed Changes

The `@automattic/composite-checkout` package has form status (loading, ready, submitting, validating, complete) and transaction status (not-started, pending, complete, redirecting, error). Conceptually these are separate things, but they are intertwined. `useTransactionStatusHandler()` updates the form state based on the transaction state (amongst other tasks). Because each state is its own React hook, this means that there are at least two render cycles when transaction status changes before the form gets updated.

Normally this is fine but it is possible it could cause race conditions, it's not very efficient, and it's prone to bugs where the two interact. A better solution would be to have both states stored in the same reducer so that they can update simultaneously. In this PR we combine both states into one while maintaining separate, unchanged, APIs for both.

Fixes https://github.com/Automattic/wp-calypso/issues/82067

Fixes https://github.com/Automattic/payments-shilling/issues/2379

Before:

https://github.com/Automattic/wp-calypso/assets/2036909/4565b12b-fc32-4121-b322-1aa59859abac 

After:

https://github.com/Automattic/wp-calypso/assets/2036909/9dad7038-e7de-43bf-8d16-b79a47458295


## Testing Instructions

Automated tests are included. To test the bug fix for https://github.com/Automattic/payments-shilling/issues/2379, do the following:

- Visit checkout with a product in your cart.
- Submit checkout with a payment method which can be very slow (eg: WeChatPay, which you can test by following the instructions in https://github.com/Automattic/wp-calypso/pull/86775).
- While the payment is pending, the form elements (eg: the buttons) will be marked as disabled.
- Unfocus the window for about a minute and then come back. This will cause the shopping cart to refresh.
- Verify that the form remains in a disabled state.
